src/fugw/solvers/utils.py

```diff
@@ -938,18 +1077,96 @@ def elementwise_prod_fact_sparse(a, b, p):
     )
 
 
+def compute_approx_kl(p, q):
+    # By convention: 0 log 0 = 0
+    entropy = torch.nan_to_num(
+        p * (p / q).log(), nan=0.0, posinf=0.0, neginf=0.0
+    ).sum()
+    return entropy
+
+
 def compute_approx_kl_sparse(p, q):
     return compute_approx_kl(p.values(), q.values())
 
 
 def compute_kl(p, q):
+    """Compute Kullback-Leibler divergence between two distributions.
+
+    Parameters
+    ----------
+    p: torch tensor
+    q: torch tensor
+        Should have the same size as p
+
+    Returns
+    -------
+    kl: float
+    """
     return compute_approx_kl(p, q) - p.sum() + q.sum()
 
 
 def compute_kl_sparse(p, q):
+    """Compute Kullback-Leibler divergence between two distributions.
+
+    Parameters
+    ----------
+    p: torch sparse CSR tensor
+    q: torch sparse CSR tensor
+        Should have the same size and sparsity mask as p
+
+    Returns
+    -------
+    kl: float
+    """
     return compute_approx_kl_sparse(p, q) - csr_sum(p) + csr_sum(q)
 
 
+def compute_l2(p, q):
+    """Compute L2 distance between two distributions.
+
+    Parameters
+    ----------
+    p: torch tensor
+    q: torch tensor
+        Should have the same size as p
+
+    Returns
+    -------
+    l2: float
+    """
+    return torch.sum((p - q) ** 2)
+
+
+def compute_l2_sparse(p, q):
+    """Compute L2 distance between two distributions.
+
+    Parameters
+    ----------
+    p: torch sparse CSR tensor
+    q: torch sparse CSR tensor
+        Should have the same size and sparsity mask as p
```

Comment:
This is not checked ?

Reply:
Indeed, this functions are not tested yet. Would be an important addition to make in a follow-up PR.